### ISSUE TITLE
nvmf/nvme: Subsystem not transition to paused when IO is outstanding during hot remove

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#1552]
Testing hot plugging under load (perf) with 20.07 plus patches from Seth (https://review.spdk.io/gerrit/c/spdk/spdk/+/3759).
When a device is detected (after it was removed), it is added back to a specific subsystem. The RPC to add the namespace to the subsystem can hang if there's an io outstanding (see trace below).
Note, there are hot adds/removes happening in quick succession which causes this issue. 

Note, Seth's patches fix another issue and are unrelated to this issue.

## Expected Behavior
RPCs that pause subsystem must not hang.

## Current Behavior

The trace below shows that when there's an outstanding IO for a subsystem, the add namespace RPC can hang. 

```
2020-08-13 15:25:03.771 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=697
2020-08-13 15:25:03.791 PDT {25422} W {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:176:Scan Did not find newly registered bdev HotInNvme88n1 in bdev list
2020-08-13 15:25:03.791 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:213:Scan Namespace HotInNvme88n1 was hot removed
2020-08-13 15:25:03.791 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:258:Scan Number of available devices: 23
2020-08-13 15:25:03.796 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=697
2020-08-13 15:25:05.652 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=662
2020-08-13 15:25:05.668 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=662
2020-08-13 15:25:06.883 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=662
2020-08-13 15:25:07.015 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=697
2020-08-13 15:25:07.057 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2020-08-13 15:25:07.117 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:22 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:22 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.117 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.117 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:21 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:21 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.117 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.117 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:20 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:20 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.117 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.117 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.117 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:19 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:19 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.118 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.118 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:18 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:18 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.118 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.118 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:17 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:17 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.118 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.118 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:16 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:16 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.118 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.118 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:15 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:07.118 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:15 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:07.118 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:07.118 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_ctrlr.c:799:nvme_ctrlr_fail ctrlr 0000:c7:00.0 in failed state.

// Here we are trying to add a newly discovered device to a subsystem. The name of the device was picked of from the bdev events.
// The device may no longer be valid and hot removed but we don't know that yet.
2020-08-13 15:25:07.122 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:187:Scan Adding HotInNvme89n1 to subsystem nqn.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
2020-08-13 15:25:07.154 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2020-08-13 15:25:07.203 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=662

// This is the corresponding RPC for adding the namespace to the subsystem
2020-08-13 15:25:07.203 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=666
2020-08-13 15:25:07.203 PDT {25422} I {nvmf_tgt_app_rpc_timeout} subsystem.c:454:nvmf_subsystem_set_state !!! mhae: nvmf_subsystem_set_state: state=3 actual_old_state=2, expected=2, subsystem->changing_state=1
2020-08-13 15:25:07.203 PDT {25422} I {nvmf_tgt_app_rpc_timeout} subsystem.c:644:spdk_nvmf_subsystem_pause !!! mhae spdk_nvmf_subsystem_pause: rc=0

// Notice that we have 1 outstandnig IO on the subsystem. This shouldn't happen if the ctrlr was failed and nvmf_complete_request callback was called.
2020-08-13 15:25:07.204 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvmf.c:1380:nvmf_poll_group_pause_subsystem !!! mhae: nvmf_poll_group_pause_subsystem: have ios: sgroup->io_outstanding=1
2020-08-13 15:25:07.298 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=697
2020-08-13 15:25:07.339 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: eal_parse_sysfs_value(): cannot open sysfs value /sys/bus/pci/devices/0000:c7:00.0/vendor
2020-08-13 15:25:07.339 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Scan for (pci) bus failed.
2020-08-13 15:25:12.669 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=697
2020-08-13 15:25:12.719 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=697
2020-08-13 15:25:14.989 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=697
2020-08-13 15:25:15.155 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2020-08-13 15:25:15.209 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=697
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:22 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:22 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:21 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:21 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:20 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:20 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:19 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:19 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:18 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:18 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:17 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:17 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:16 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:16 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_pcie.c:1444:nvme_pcie_qpair_abort_trackers aborting outstanding command
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:250:nvme_admin_qpair_print_command ASYNC EVENT REQUEST (0c) qid:0 cid:15 nsid:0 cdw10:00000000 cdw11:00000000
2020-08-13 15:25:15.222 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_qpair.c:454:spdk_nvme_print_completion ABORTED - BY REQUEST (00/07) qid:0 cid:15 cdw0:0 sqhd:0000 p:0 m:0 dnr:0
2020-08-13 15:25:15.222 PDT {25422} W {nvmf_tgt_app_rpc_timeout} bdev_nvme.c:1342:aer_cb AER request execute failed
2020-08-13 15:25:15.222 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_ctrlr.c:799:nvme_ctrlr_fail ctrlr 0000:c7:00.0 in failed state.
2020-08-13 15:25:15.258 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Unexpected device scan at /sys/bus/pci/devices/0000:c7:00.0/driver!
2020-08-13 15:25:15.441 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: eal_parse_sysfs_value(): cannot open sysfs value /sys/bus/pci/devices/0000:c7:00.0/vendor
2020-08-13 15:25:15.441 PDT {25422} E {nvmf_tgt_app_rpc_timeout} DPDK:0: EAL: Scan for (pci) bus failed.

// I added code to timeout a RPC after 10s. Getting to this point means that _nvmf_complete_request wasn't called for that outstanding IO to transition the subsystem into paused stated.
2020-08-13 15:25:18.001 PDT {25422} E {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:196:Scan Unable to add HotInNvme89n1 to subsystem nqn.2019-20.com.hpe:51402ec01492a822.0.bay07 status=CMD_FAILED:
 !!! Error during waiting for server response, rc=-110

2020-08-13 15:25:18.043 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=233
2020-08-13 15:25:18.043 PDT {25422} I {nvmf_tgt_app_rpc_timeout} subsystem.c:644:spdk_nvmf_subsystem_pause !!! mhae spdk_nvmf_subsystem_pause: rc=-16
2020-08-13 15:25:18.065 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:201:Scan Tried removing failed namespace, status=CMD_FAILED: Unexpected response from server: {"code":-32603,"message":"subsystem sy, retry later.
n"}}

2020-08-13 15:25:18.065 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:213:Scan Namespace HotInNvme89n1 was hot removed
2020-08-13 15:25:18.134 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:241:jsonrpc_server_accept !!! mhae: Client connected: sockfd=662
2020-08-13 15:25:18.146 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=233
2020-08-13 15:25:18.155 PDT {25422} W {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:176:Scan Did not find newly registered bdev HotInNvme90n1 in bdev list
2020-08-13 15:25:18.155 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:213:Scan Namespace HotInNvme90n1 was hot removed
2020-08-13 15:25:18.155 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:233:Scan Device at 0000:c7:00.0 is missing or setting user mode driver failed
2020-08-13 15:25:18.155 PDT {25422} I {nvmf_tgt_app_rpc_timeout} nvme_device_discovery.cc:258:Scan Number of available devices: 23
2020-08-13 15:25:18.240 PDT {25422} I {nvmf_tgt_app_rpc_timeout} jsonrpc_server_tcp.c:132:jsonrpc_server_conn_close !!! mhae: Client disconnected: sockfd=662
```
In the above scenario, it seems that `_nvmf_request_complete` is not called for the outstanding IO since this function is responsible for transitioning a pausing subsystem into paused state:
```
_nvmf_request_complete:
~~~
if (sgroup && !is_aer) {
		assert(sgroup->io_outstanding > 0);
		sgroup->io_outstanding--;
		if (sgroup->state == SPDK_NVMF_SUBSYSTEM_PAUSING &&
		    sgroup->io_outstanding == 0) {
			sgroup->state = SPDK_NVMF_SUBSYSTEM_PAUSED;
			sgroup->cb_fn(sgroup->cb_arg, 0);
		}
	}
~~~
```
The place where the state is set is in:
```
nvmf_poll_group_pause_subsystem:
~~~
if (sgroup->io_outstanding > 0) {
		SPDK_NOTICELOG("!!! nvmf_poll_group_pause_subsystem: have ios: sgroup->io_outstanding=%lu\n", sgroup->io_outstanding);
		sgroup->cb_fn = cb_fn;
		sgroup->cb_arg = cb_arg;
		return;
	}
~~~
```

## Possible Solution

It seems to me that we may not clean up a failed NVMe ctrlr correctly since this should trigger the callback to _nvmf_complete_request.

## Steps to Reproduce
Somewhat tricky ... multiple subsystems with NVMe drives (one in each subsystem). Perf load to all subsystem. Hot plug script that removes/adds drives in quick succession in a loop.

## Context (Environment including OS version, SPDK version, etc.)
20.07

**Comment by mhae** [comment]
I think I found a clue. In my testing, the path for SPDK_NVME_ASYNC_EVENT_TYPE_NOTICE was triggered without decrementing the io_outstanding.
```
tatic int
nvmf_ctrlr_async_event_request(struct spdk_nvmf_request *req)
{
	struct spdk_nvmf_ctrlr *ctrlr = req->qpair->ctrlr;
	struct spdk_nvme_cpl *rsp = &req->rsp->nvme_cpl;
	struct spdk_nvmf_subsystem_poll_group *sgroup;

	SPDK_DEBUGLOG(SPDK_LOG_NVMF, "Async Event Request\n");

	/* Four asynchronous events are supported for now */
	if (ctrlr->nr_aer_reqs >= NVMF_MAX_ASYNC_EVENTS) {
		SPDK_DEBUGLOG(SPDK_LOG_NVMF, "AERL exceeded\n");
		rsp->status.sct = SPDK_NVME_SCT_COMMAND_SPECIFIC;
		rsp->status.sc = SPDK_NVME_SC_ASYNC_EVENT_REQUEST_LIMIT_EXCEEDED;
		return SPDK_NVMF_REQUEST_EXEC_STATUS_COMPLETE;
	}

	if (ctrlr->notice_event.bits.async_event_type ==
	    SPDK_NVME_ASYNC_EVENT_TYPE_NOTICE) {
		rsp->cdw0 = ctrlr->notice_event.raw;
		ctrlr->notice_event.raw = 0;
		return SPDK_NVMF_REQUEST_EXEC_STATUS_COMPLETE;
	}

	if (ctrlr->reservation_event.bits.async_event_type ==
	    SPDK_NVME_ASYNC_EVENT_TYPE_IO) {
		rsp->cdw0 = ctrlr->reservation_event.raw;
		ctrlr->reservation_event.raw = 0;
		return SPDK_NVMF_REQUEST_EXEC_STATUS_COMPLETE;
	}

	/* AER cmd is an exception */
	sgroup = &req->qpair->group->sgroups[ctrlr->subsys->id];
	assert(sgroup != NULL);
	sgroup->io_outstanding--;

	ctrlr->aer_req[ctrlr->nr_aer_reqs++] = req;
	return SPDK_NVMF_REQUEST_EXEC_STATUS_ASYNCHRONOUS;
}```

The SPDK_NVME_ASYNC_EVENT_TYPE_NOTICE is set in nvmf_ctrlr_async_event_ns_notice when nr_aers_reqs=0.

**Comment by Seth5141** [comment]
Just posting mhae's change here for informational purposes. https://review.spdk.io/gerrit/c/spdk/spdk/+/3854

**Comment by tomzawadzki** [comment]
Patch https://review.spdk.io/gerrit/c/spdk/spdk/+/3854 is now merged, closing this issue. Thank you @mhae !

[FYI to automatically close GH issues from patches, please use [this](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) guide. So replacing `Fixes: 1552` with `Fixes #1552` would do the trick.]
